jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormData.java

```diff
@@ -295,9 +403,47 @@ public Parser(String boundary, MultiPartCompliance multiPartCompliance, Complian
             parser = new MultiPart.Parser(Objects.requireNonNull(boundary), compliance, listener);
         }
 
+        public void parse(Content.Source content, Promise<Parts> immediate, Invocable.InvocablePromise<Parts> future)
```

Comment:
Looking at this implementation (and the future ones we could have) I wonder what's the usefulness of `Promise<Parts> immediate` as figuring out if that one should be called or the future one is always going to be racy?

Reply:
It is a little bit racy, but it is essentially an optimization to avoid unnecessary executes, so if we lose the race, we typically just execute instead of calling directly when the results are available.

Reply:
@lorban also note that if we lose the race, the invocation type of the future promise would have been used for the demand callback, so we are good.  That is why we need both the immediate and future promises, as whilst we attempt the read and parse, we are using the future promises invocation type, but if we manage to completely read the content, then we can call the immediate promise as we were not actually invoked during the reading/parsing and are still in the original calling thread.